stetho/src/main/java/com/facebook/stetho/inspector/elements/android/ApplicationDescriptor.java

```diff
@@ -0,0 +1,155 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+package com.facebook.stetho.inspector.elements.android;
+
+import android.app.Activity;
+import android.app.Application;
+import android.os.Build;
+import android.os.Bundle;
+
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.elements.ChainedDescriptor;
+import com.facebook.stetho.inspector.elements.NodeType;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final class ApplicationDescriptor extends ChainedDescriptor<Application> {
+  private final Map<Application, ElementContext> mElementToContextMap =
+      Collections.synchronizedMap(new HashMap<Application, ElementContext>());
+
+  private ElementContext getContext(Application element) {
+    return mElementToContextMap.get(element);
+  }
+
+  @Override
+  protected void onHook(Application element) {
+    ElementContext context = new ElementContext();
+    context.hook(element);
+    mElementToContextMap.put(element, context);
+  }
+
+  @Override
+  protected void onUnhook(Application element) {
+    ElementContext context = mElementToContextMap.remove(element);
+    context.unhook();
+  }
+
+  @Override
+  protected NodeType onGetNodeType(Application element) {
+    return NodeType.DOCUMENT_NODE;
+  }
+
+  @Override
+  protected int onGetChildCount(Application element) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().size();
+  }
+
+  @Override
+  protected Object onGetChildAt(Application element, int index) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().get(index);
+  }
+
+  private final class ElementContext {
+    private Application mElement;
+    private Application.ActivityLifecycleCallbacks mCallbacks;
+    private List<Activity> mActivities;
+
+    public void hook(Application element) {
+      mElement = Util.throwIfNull(element);
+      mActivities = new ArrayList<>();
+
+      // TODO: tree diffing will remove the need to even worry about installing this callback,
+      //       which will then allow ~realtime updates for pre-ICS. for now, pre-ICS will just
+      //       have to close and re-open the inspector whenever the activity changes.
+      //       (this does assume that the hack below for getAllActivitiesHack() works on pre-ICS)
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
```

Comment:
Never write this imperatively without a class guarding the implementation.  Doing so causes significant performance problems when the verifier runs on this code -- it sprays logcat with warnings about having to rewrite opcodes on the fly.

Always produce a compat class using the pattern you have for `ViewHighlighter` which is designed to isolate the verifier from the illegal code on older versions.